k2/csrc/fsa_algo_test.cu

```diff
@@ -1053,4 +1054,67 @@ TEST(FsaAlgo, TestRemoveEplsionSelfRandomFsas) {
   }
 }
 
+TEST(FsaAlgo, TestReplaceFsaA) {
+  for (const ContextPtr &c : {GetCpuContext(), GetCudaContext()}) {
+    FsaVec src = FsaVec(" [ [ [ 0 1 1 1.0 0 2 2 2.0 ] [ 1 3 -1 0.0 ] "
+                          "     [ 2 3 -1 0 ] [ ] ] ]").To(c);
+    FsaVec index = FsaVec("[ [ [ ] [ ] ] ]").To(c);
+    Array1<int32_t> arc_src_map, arc_index_map;
+    FsaVec replaced_fsa =
+        ReplaceFsa(src, index, 0, &arc_src_map, &arc_index_map);
+
+    K2_CHECK(Equal(replaced_fsa, index));
+  }
+}
+
+TEST(FsaAlgo, TestReplaceFsaB) {
+  for (const ContextPtr &c : {GetCpuContext(), GetCudaContext()}) {
+    FsaVec src = FsaVec("[ [ [ ] [ ] ] ]").To(c);
+    FsaVec index = FsaVec(" [ [ [ 0 1 1 1.0 0 2 2 2.0 ] [ 1 3 -1 0.0 ] "
+                          "     [ 2 3 -1 0 ] [ ] ] ]").To(c);
+    Array1<int32_t> arc_src_map, arc_index_map;
+    FsaVec replaced_fsa =
+        ReplaceFsa(src, index, 0, &arc_src_map, &arc_index_map);
+
+    Array1<int32_t> arc_src_map_ref(c, "[ -1 -1 -1 -1 ]"),
+                    arc_index_map_ref(c, "[ 0 1 2 3 ]");
+
+    K2_CHECK(Equal(replaced_fsa, index));
+    K2_CHECK(Equal(arc_src_map, arc_src_map_ref));
+    K2_CHECK(Equal(arc_index_map, arc_index_map_ref));
+  }
+}
+
+TEST(FsaAlgo, TestReplaceFsaC) {
+  for (const ContextPtr &c : {GetCpuContext(), GetCudaContext()}) {
+    FsaVec src = FsaVec("[ [ [ 0 1 11 11.0 ] [ 1 2 12 12.0 ] "
+                        "    [ 2 3 -1 0.0] [ ] ] "
+                        "  [ [ 0 1 21 21.0 0 2 22 22.0 ] "
+                        "    [ 1 3 -1 0.0 ] [2 3 -1 0.0] [ ] ] "
+                        "  [ [ 0 1 31 31.0 0 2 33 33.0 ] "
+                        "    [ 1 2 32 32.0 ] [ 2 3 -1 0.0 ] [ ] ] ]").To(c);
+    FsaVec index = FsaVec(" [ [ [ 0 1 1 1.0 0 2 2 2.0 ] [ 1 3 3 3.0 ] "
+                          "     [ 2 4 4 4.0 2 5 -1 0.0 ] [ 3 5 -1 0.0 ] "
+                          "     [ 4 5 -1 0 ] [ ] ] ]").To(c);
+    Array1<int32_t> arc_src_map, arc_index_map;
+    FsaVec replaced_fsa =
+        ReplaceFsa(src, index, 0, &arc_src_map, &arc_index_map);
```

Comment:
Are you ever testing this where the 3rd arg is nonzero?

Reply:
Also: in order to properly test the semantics of this, especially with randomized FSAs, it would be more ideal to check its equivalence with composition.  This operation is intended as a cheaper replacement for composition; you should be able to figure out the details of this.  However, the equivalence may only hold after removing epsilons in the output.  (This may be easier to test at the python level).

Reply:
> Are you ever testing this where the 3rd arg is nonzero?

I test it in python level, will add the test case here.


> Also: in order to properly test the semantics of this, especially with randomized FSAs, it would be more ideal to check its equivalence with composition. This operation is intended as a cheaper replacement for composition; you should be able to figure out the details of this. However, the equivalence may only hold after removing epsilons in the output. (This may be easier to test at the python level).

Will figure out the relationship of replace and composition and try to add test case.

Reply:
@danpovey  Sorry, I still haven't figured out how replace act as a replacement of composition, would you please show me more details, or, some simple test cases, thanks!

Reply:
Consider what happens if you compose the "index" FSA with an FSA, say f, that has the following structure:
the start-state of f has an arc directly to final-state with -1 label, and:
 For labels `symbol_range_begin <= s < src.Dim0()`:
    There is an arc with ilabel == s and olabel == 0, from state 0 of f to a state in f that corresponds 
  to the start-state of the
  FSA f_s = src[s-symbol_range_begin].  The FSA f_s has its labels as the olabels of f, with epsilons as the
  ilabels; and its final-state is identified with the start-state of f.  Other states in f_s are just copied to new 
  states in f.  The -1 labels in f_s would be converted to 0 olabels in f.
To get epsilons treated correctly in composition assuming treat_epsilons_specially=false, we'd have to
add epsilon self-loops to `src` before doing the composition.  (Note: things wouldn't work quite right
if symbol_range_begin <= 0). 
Anyway, the result of composing src with f should be equivalent to the result of the `replace` operation.

Reply:
A simple example to show whether I understood right.

For the replace operation, the following `index` and `src`

|index|src{0,1}|
|--|--|
|![image](https://user-images.githubusercontent.com/11765074/124244472-8e3b0a80-db51-11eb-932b-ab74ee2068a4.png)|![image](https://user-images.githubusercontent.com/11765074/124244421-7fecee80-db51-11eb-90ce-f920bb09b27b.png)|
||![image](https://user-images.githubusercontent.com/11765074/124244446-85e2cf80-db51-11eb-9e31-fa5da5786984.png)|

with `symbol_range_begin=1` the output is
![image](https://user-images.githubusercontent.com/11765074/124244113-2d133700-db51-11eb-861e-64dd552417d1.png)

For the compose operation, I construct fsa `f` as the explanation above:

|inverted index|f|
|--|--|
|![image](https://user-images.githubusercontent.com/11765074/124248340-533ad600-db55-11eb-826f-67f26b6e4ff6.png)|![image](https://user-images.githubusercontent.com/11765074/124245699-d0188080-db52-11eb-940f-4986b7490923.png)|

then I did `k2.invert(k2.compose(k2.invert(index), f))`, the output is

![image](https://user-images.githubusercontent.com/11765074/124245486-9c3d5b00-db52-11eb-8e5c-56c66e4a0ec7.png)

the only difference is the olabel of the arc points to the final state.

So, if I want to check the equivalence between replace and composition, I should construct a fsa `f` from src, then compare the result of `k2.replace(index, src)` and `k2.invert(k2.compose(k2.invert(index), f))`

Reply:
This looks plausible although I haven't checked carefully.
What I find strange is that your index and src have olabels.  olabels should not be needed for doing the indexing or the comparison with the composed result.
Ultimately you should be able to remove the olabels and compare it as a simple FSA.